File: frontend/services/milestone.ts
import { api } from './api'

export interface Milestone {
  id: string
  project_id: string
  title: string
  description: string | null
  due_date: string | null
  status: string // 'Planned', 'Active', 'Completed'
  is_archived: boolean
  progress_percentage: number
  total_tasks: number
  completed_tasks: number
  created_at: string
  updated_at: string
}

export interface MilestoneStats {
  milestone_id: string
  title: string
  status: string
  total_tasks: number
  completed_tasks: number
  remaining_tasks: number
  completion_percentage: number
  due_date: string | null
  is_archived: boolean
  timeline: { label: string; date: string; status: string }[]
  tasks: any[]
}

export const milestoneService = {
  async getMilestones(projectId: string, filters?: Record<string, any>): Promise<Milestone[]> {
    const params = { project_id: projectId, ...filters }
    const response = await api.get<Milestone[]>('/milestones', { params })
    return response.data
  },

  async getMilestoneById(id: string): Promise<Milestone> {
    const response = await api.get<Milestone>(`/milestones/${id}`)
    return response.data
  },

  async createMilestone(data: {
    project_id: string
    title: string
    description?: string
    due_date?: string
    status?: string
  }): Promise<Milestone> {
    const response = await api.post<Milestone>('/milestones', data)
    return response.data
  },

  async updateMilestone(id: string, data: Partial<Milestone>): Promise<Milestone> {
    const response = await api.patch<Milestone>(`/milestones/${id}`, data)
    return response.data
  },

  async archiveMilestone(id: string): Promise<Milestone> {
    const response = await api.post<Milestone>(`/milestones/${id}/archive`)
    return response.data
  },

  async deleteMilestone(id: string): Promise<void> {
    await api.delete(`/milestones/${id}`)
  },

  async assignTasksToMilestone(id: string, taskIds: string[]): Promise<any> {
    const response = await api.post<any>(`/milestones/${id}/tasks`, { task_ids: taskIds })
    return response.data
  },

  async removeTaskFromMilestone(id: string, taskId: string): Promise<any> {
    const response = await api.delete<any>(`/milestones/${id}/tasks/${taskId}`)
    return response.data
  },

  async getMilestoneStats(id: string): Promise<MilestoneStats> {
    const response = await api.get<MilestoneStats>(`/milestones/${id}/stats`)
    return response.data
  },

  async getUpcomingMilestones(limit = 5): Promise<MilestoneStats[]> {
    const response = await api.get<MilestoneStats[]>('/milestones/upcoming', { params: { limit } })
    return response.data
  }
}

File: backend/app/schemas/milestone.py
import uuid
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from app.schemas.task import TaskResponse

class MilestoneBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1000)
    due_date: Optional[datetime] = None
    status: str = Field("Planned", max_length=50) # 'Planned', 'Active', 'Completed'
    is_archived: bool = False

class MilestoneCreate(MilestoneBase):
    project_id: uuid.UUID

class MilestoneUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1000)
    due_date: Optional[datetime] = None
    status: Optional[str] = Field(None, max_length=50)
    is_archived: Optional[bool] = None

class MilestoneTaskAssignment(BaseModel):
    task_ids: List[uuid.UUID]

class MilestoneResponse(MilestoneBase):
    id: uuid.UUID
    project_id: uuid.UUID
    progress_percentage: float = 0
    total_tasks: int = 0
    completed_tasks: int = 0
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

class MilestoneStatsResponse(BaseModel):
    milestone_id: uuid.UUID
    title: str
    description: Optional[str] = None
    status: str
    total_tasks: int
    completed_tasks: int
    remaining_tasks: int
    completion_percentage: float
    due_date: Optional[datetime] = None
    is_archived: bool
    timeline: List[dict]
    tasks: List[TaskResponse] = []
